src/Swarm/Util.hs

```diff
@@ -83,27 +94,38 @@ maximum0 :: (Num a, Ord a) => [a] -> a
 maximum0 [] = 0
 maximum0 xs = maximum xs
 
--- | Safely attempt to read a file, returning @Nothing@ if the file
---   does not exist.  \"Safely\" should be read in scare quotes here,
---   since /e.g./ we do nothing to guard against the possibility of a
---   race condition where the file is deleted after the existence
---   check but before trying to read it.  But it's not like we're
---   worried about security or anything here.
-readFileMay :: FilePath -> IO (Maybe String)
-readFileMay file = do
-  b <- doesFileExist file
-  case b of
-    False -> return Nothing
-    True -> Just <$> readFile file
-
 -- | Take the successor of an 'Enum' type, wrapping around when it
 --   reaches the end.
 cycleEnum :: (Eq e, Enum e, Bounded e) => e -> e
 cycleEnum e
   | e == maxBound = minBound
   | otherwise = succ e
 
---------------------------------------------------
+------------------------------------------------------------
+-- Directory stuff
+
+-- | Safely attempt to read a file.
+readFileMay :: FilePath -> IO (Maybe String)
+readFileMay = catchIO . readFile
+
+-- | Safely attempt to (efficiently) read a file.
+readFileMayT :: FilePath -> IO (Maybe Text)
+readFileMayT = catchIO . T.readFile
+
+-- | Turns any IO error into Nothing.
+catchIO :: IO a -> IO (Maybe a)
+catchIO act = (Just <$> act) `catchIOError` (\_ -> return Nothing)
+
+-- | Get path to swarm history, optionally creating necessary
+--   directories. This could fail if user has bad permissions
+--   on his own $HOME or $XDG_DATA_HOME which is unlikely.
+getSwarmHistoryPath :: Bool -> IO FilePath
```

Comment:
In the future we may want to put other things in the swarm data dir (e.g. save files).  How about generalizing this to `getSwarmPath` which takes the name of the wanted file as a `FilePath` argument, instead of hard-coding the file name `"history"`?